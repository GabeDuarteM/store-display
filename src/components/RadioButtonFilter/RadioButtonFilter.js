import React from 'react'
import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import RadioButtonFilterItem from '../RadioButtonFilterItem'

export const styles = {
  radioButtonFilter: {
    display: 'flex',
    userSelect: 'none',
  },
}

const RadioButtonFilter = ({ classes, options, activeOption, onChange }) => (
  <div className={classes.radioButtonFilter}>
    {options.map(elem => (
      <RadioButtonFilterItem
        key={elem.id}
        onChange={() => onChange(elem.id)}
        text={elem.text}
        active={elem.id === activeOption}
      />
    ))}
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Array of items that will be displayed inside the RadioButton */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /**
   * Gets called when the user clicks on the radiobutton
   *
   * @param {string} id Id of the option that was clicked
   */
  onChange: PropTypes.func.isRequired,
  /** The active filter */
  activeOption: PropTypes.string,
}

export const defaultProps = {
  activeOption: null,
}

RadioButtonFilter.propTypes = propTypes
RadioButtonFilter.defaultProps = defaultProps

export default injectSheet(styles)(RadioButtonFilter)
