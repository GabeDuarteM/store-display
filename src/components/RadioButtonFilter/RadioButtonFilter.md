```js
initialState = {
  options: [{ id: 'woman', text: 'Woman' }, { id: 'man', text: 'Man' }],
  activeOption: 'woman',
}
;<RadioButtonFilter
  onChange={id => {
    setState({
      activeOption: id,
    })
  }}
  options={state.options}
  activeOption={state.activeOption}
/>
```
