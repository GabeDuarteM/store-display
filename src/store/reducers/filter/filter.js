import {
  FILTER__CHANGE_GENDER,
  FILTER__CHANGE_SORT,
  FILTER__LOAD_MORE_PRODUCTS,
} from './filterActions'

const returnInitialState = () => ({
  gender: 'woman',
  sort: 'price-asc',
  productListLength: 9,
})

const filter = (state = returnInitialState(), action = {}) => {
  switch (action.type) {
    case FILTER__CHANGE_GENDER:
      return {
        ...state,
        gender: action.payload,
        productListLength: 9,
      }
    case FILTER__CHANGE_SORT:
      return {
        ...state,
        sort: action.payload,
      }
    case FILTER__LOAD_MORE_PRODUCTS:
      return {
        ...state,
        productListLength: state.productListLength + 9,
      }
    default:
      return state
  }
}

export default filter
