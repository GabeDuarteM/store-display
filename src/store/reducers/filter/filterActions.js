export const FILTER__CHANGE_GENDER = 'FILTER/CHANGE_GENDER'
export const FILTER__CHANGE_SORT = 'FILTER/CHANGE_SORT'
export const FILTER__LOAD_MORE_PRODUCTS = 'FILTER/LOAD_MORE_PRODUCTS'

export const filterChangeGender = gender => ({
  type: FILTER__CHANGE_GENDER,
  payload: gender,
})

export const filterChangeSort = sort => ({
  type: FILTER__CHANGE_SORT,
  payload: sort,
})

export const filterLoadMoreProducts = () => ({
  type: FILTER__LOAD_MORE_PRODUCTS,
})
