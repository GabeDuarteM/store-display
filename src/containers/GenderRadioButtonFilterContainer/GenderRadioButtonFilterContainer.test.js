import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import GenderRadioButtonFilterContainerConnected, {
  GenderRadioButtonFilterContainer,
} from './GenderRadioButtonFilterContainer'
import RadioButtonFilter from '../../components/RadioButtonFilter/RadioButtonFilter'
import { filterChangeGender as filterChangeGenderAction } from '../../store/reducers/filter/filterActions'

const createWrapper = (func, optionalProps) =>
  func(
    <GenderRadioButtonFilterContainer
      activeOption="woman"
      filterChangeGender={() => {}}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <GenderRadioButtonFilterContainer />', () => {
  it('should render without crashing', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it('should have a RadioButtonFilter component', () => {
    const wrapper = createWrapperShallow()
    const hasRadioButtonFilter = wrapper.find(RadioButtonFilter).exists()

    expect(hasRadioButtonFilter).toBe(true)
  })

  it('should render RadioButtonFilter with the correct props', () => {
    const genders = [{ id: 'woman', text: 'Woman' }, { id: 'man', text: 'Man' }]
    const activeOption = 'man'
    const filterChangeGender = jest.fn()

    const wrapper = createWrapperShallow({ activeOption, filterChangeGender })
    const gendersProp = wrapper.dive().prop('options')
    const activeOptionProp = wrapper.dive().prop('activeOption')
    const filterChangeGenderProp = wrapper.dive().prop('onChange')

    expect(gendersProp).toEqual(genders)
    expect(activeOptionProp).toBe(activeOption)
    expect(filterChangeGenderProp).toBe(filterChangeGender)
  })

  it('should have a connected component rendering the props correctly', () => {
    const activeOption = 'man'
    const store = {
      getState: () => ({
        filter: { gender: activeOption },
      }),
      dispatch: jest.fn(),
      subscribe: () => {},
    }

    const component = shallow(
      <GenderRadioButtonFilterContainerConnected store={store} />,
    ).find(GenderRadioButtonFilterContainer)
    const activeOptionProp = component.prop('activeOption')
    const filterChangeGenderProp = component.prop('filterChangeGender')
    filterChangeGenderProp(activeOption)

    expect(activeOptionProp).toBe(activeOption)
    expect(store.dispatch).toHaveBeenCalled()
    expect(store.dispatch).toHaveBeenCalledWith(
      filterChangeGenderAction(activeOption),
    )
  })

  it('should have a connected component that scrolls to the top when filterChangeGender is fired', () => {
    const activeOption = 'man'
    const store = {
      getState: () => ({
        filter: { gender: activeOption },
      }),
      dispatch: jest.fn(),
      subscribe: () => {},
    }

    const component = shallow(
      <GenderRadioButtonFilterContainerConnected store={store} />,
    ).find(GenderRadioButtonFilterContainer)

    const filterChangeGenderProp = component.prop('filterChangeGender')
    window.scrollTo.mockClear()
    filterChangeGenderProp(activeOption)

    expect(window.scrollTo).toHaveBeenCalledTimes(1)
    expect(window.scrollTo).toHaveBeenLastCalledWith(0, 0)
  })
})
