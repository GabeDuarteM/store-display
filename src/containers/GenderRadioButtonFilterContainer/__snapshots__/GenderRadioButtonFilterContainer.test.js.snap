// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <GenderRadioButtonFilterContainer /> should match component's snapshot 1`] = `
<RadioButtonFilter
  activeOption="woman"
  classes={
    Object {
      "radioButtonFilter": "RadioButtonFilter-radioButtonFilter",
    }
  }
  onChange={[Function]}
  options={
    Array [
      Object {
        "id": "woman",
        "text": "Woman",
      },
      Object {
        "id": "man",
        "text": "Man",
      },
    ]
  }
/>
`;
